modules/4-graphql.livemd

```diff
@@ -2,89 +2,160 @@
 
-*TODO: Write Introduction*
+Since GraphQL can be implemented as a component of an application's API, there are security issues common to all APIs present, as well as concerns related to characteristics of the query language itself.  This module will highlight several security issues associated with GraphQL and recommendations for how to address them.
 
 ## Table of Contents
 
 * [Disabling Introspection](#disabling-introspection)
 * [Error Disclosure](#error-disclosure)
-* [Resource Exhaustion](#resource-disclosure)
-  * [Cost Theory](#cost-theory)
+* [Resource Exhaustion](#resource-exhaustion)
+* [Cost Theory](#cost-theory)
 
 ## Disabling Introspection
 
 ### Description
 
-*TODO: Write Description*
+Introspection queries are a way of enumerating a particular GraphQL implementation to discover details about the queries supported, data types available, and other information. This includes mutation names, fields specific to an organization or dataset, query parameters, and types of objects in the data source.  Obtaining this information can help a user, including a malicious one, deduce and discover specifics about the data being stored. 
+
+If you are familiar with databases, this is similar to gathering info on the [database schema]( https://en.wikipedia.org/wiki/Database_schema) that includes information about table names, fields, database, structure etc.
+
+Malicious actors in their information gathering/reconnaissnce efforts can leverage this information as they look for ways to attack your application and construct malicious queries and requests to compromise data.
+
+Excessive Data Exposure is number 3 on OWASP's API Security Top 2019 and APIs with this issue return too much and/or sensitive information in response to incoming requests and queries.  Although it provides a useful function for GraphQL developers, the information returned by introspection can help facilitate attack. 
 
 ### Prevention
 
-*TODO: Write Prevention*
+Best practice per OWASP is to limit access, including following least privilege, to introspection queries if it is not possible to completely disable it. Please see references for more details.
 
-### <span style="color:blue;">Example</span> / <span style="color:red;">Quiz</span>
+### References
+https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/12-API_Testing/01-Testing_GraphQL
+
+https://cybervelia.com/?p=736
+
+### <span style="color:red;">Quiz</span>
+
+**Which of the OWASP API Security Top 10 2019 issues does disabling introspection queries address?**
```

Comment:
I'm not sure I love the idea of requiring independent research being a barrier for completing this quiz question. If we're introducing Vigil as a recommended, Elixir specific, prevention method - we should consider making a Quiz/Example that incorporates that library rather than forcing more reading of OWASP resources

Reply:
Amended my original comment since I noticed you lead the reader to the answer in the above section, my mistake! But my point of reframing the quiz/example around Vigil still stands since it's a bit more "hands-on" and Elixir specific

Reply:
Updated disabling introspection section and added reference to Vigil